ocaml/xenopsd/xc/mem_stats.ml

```diff
@@ -0,0 +1,391 @@
+    ]
+
+  let memory_allocation_approx_expensive (gc, _control) =
+    [
+      (* see https://github.com/ocaml/ocaml/blob/trunk/stdlib/gc.mli#L50-L59, without running a major
+          cycle the live_words may overestimate the actual live words, "live" just means "not currently
+         known to be collectible"*)
+      gc.live_words
+      |> maybe_words "ocaml_maybe_live"
+           "OCaml memory not currently known to be collectable by %s"
+    ; gc.free_words |> maybe_words "ocaml_free" "OCaml memory available to %s"
+    ]
+end
+
+module Derived = struct
+  open SimpleMetrics
+
+  let memextra_kib =
+    kib "mem_extra" "Total amount of non-OCaml and non-stack memory used by %s"
+
+  let ulimit_stack =
+    let ic = Unix.open_process_in "ulimit -s" in
+    let r = ic |> input_line |> Int64.of_string in
+    close_in_noerr ic ; Int64.to_float r
+
+  let memextra_kib stats (gc_stat, control) =
+    let ocaml_total_kib = GcStat.ocaml_total gc_stat control |> to_float in
+    let vmdata = Proc.find Proc.Fields.vmdata stats in
+    let vmpte = Proc.find Proc.Fields.vmpte stats in
+    let threads = Proc.find Proc.Fields.threads stats in
+    (* Each thread, except the main one will allocate 'ulimit -s' worth of VmData.
+       This won't immediately show up in VmRss until actually used (e.g. by a deep calltree due to recursion) *)
+    vmdata -. vmpte -. (ulimit_stack *. (threads -. 1.)) -. ocaml_total_kib
+    |> ds_float
+    |> memextra_kib
+end
+
+let observe_stats l =
+  let names = ListLabels.rev_map l ~f:(fun ds -> ds.Ds.ds_name) in
+  let values =
+    ListLabels.rev_map l ~f:(fun ds ->
+        let f =
+          match ds.Ds.ds_value with
+          | Rrd.VT_Int64 i ->
+              Int64.to_float i
+          | Rrd.VT_Float f ->
+              f
+          | Rrd.VT_Unknown ->
+              nan
+        in
+        ds.Ds.ds_pdp_transform_function f |> Printf.sprintf "%.0f"
+    )
+  in
+  D.debug "stats header: %s" (String.concat "," names) ;
+  D.debug "stats values: %s" (String.concat "," values)
+
+let generate_expensive_stats =
+  let generate_dss () =
+    let stat = Gc.stat () in
+    let gc_control = Gc.get () in
+    let rss = Proc.Fields.smaps_rollup () |> Proc.to_list in
+    let gcstat = GcStat.memory_allocation_approx_expensive (stat, gc_control) in
+    List.rev_append rss gcstat
+  in
+  SlowReporter.iter_of_fold (SlowReporter.report ~interval_s:150. ~generate_dss)
+
+let generate_stats () =
+  let status = Proc.Fields.status () in
+  let gc_stat = Gc.quick_stat () in
+  let gc_control = Gc.get () in
+  let derived = Derived.memextra_kib status (gc_stat, gc_control) in
+  let gcstat = GcStat.memory_allocation_precise (gc_stat, gc_control) in
+  let is_slow, slow_stats = generate_expensive_stats () in
+  let stats =
+    derived :: List.concat [gcstat; Proc.to_list status; slow_stats]
+  in
+  if is_slow then
+    observe_stats stats ;
+  stats |> List.rev_map (fun x -> (Rrd.Host, x))
+
+let generate_stats () =
```

Comment:
emm, two functions with the same name, one shadowing another? maybe define an inner function?

Reply:
Renamed one of them to have an '_exn' suffix.